internal/bsr/bsr.go

```diff
@@ -237,6 +238,145 @@ func OpenSession(ctx context.Context, sessionRecordingId string, f storage.FS, k
+		// Get connection id
+		connKey := connId
+		lastDotIndex := strings.LastIndex(connId, ".connection")
+		if lastDotIndex == -1 {
+			validation.Valid = false
+			return validation, fmt.Errorf("%s: malformed BSR for: %s", op, connId)
+		}
+		connKey = connId[:lastDotIndex]
+
+		// Open current connection
+		connection, err := session.OpenConnection(ctx, connKey)
+		if err != nil {
+			validation.Valid = false
+			return validation, fmt.Errorf("%s: failed to retrieve connection for %s: %w", op, connId, err)
+		}
+
+		connectionContainerValidation := validation.ValidateContainer(ctx, ConnectionContainer, connection.container, connection.Meta.Id)
+		sessionContainerValidation.SubContainers = append(sessionContainerValidation.SubContainers, connectionContainerValidation)
+		if err != nil {
+			validation.Valid = false
+			continue
+		}
+
+		// Validate all channels under current connection
+		for chId := range connection.Meta.channels {
+			// Get channel id
+			chKey := chId
+			lastDotIndex := strings.LastIndex(chId, ".channel")
+			if lastDotIndex == -1 {
+				validation.Valid = false
+				return validation, fmt.Errorf("%s: malformed BSR for: %s", op, chId)
+			}
+			chKey = chId[:lastDotIndex]
+
+			// Open current connection
+			channel, err := connection.OpenChannel(ctx, chKey)
+			if err != nil {
+				validation.Valid = false
+				return validation, fmt.Errorf("%s: failed to retrieve channel for %s: %w", op, chId, err)
+			}
+
+			channelContainerValidation := validation.ValidateContainer(ctx, ChannelContainer, channel.container, channel.Meta.Id)
+			connectionContainerValidation.SubContainers = append(connectionContainerValidation.SubContainers, channelContainerValidation)
+			if err != nil {
+				validation.Valid = false
+				continue
+			}
+		}
+	}
+
+	return validation, nil
+}
+
+// ValidateContainer validates the checksums of all files in a container
+func (v *Validation) ValidateContainer(ctx context.Context, ct ContainerType, c *container, name string) *ContainerValidation {
+	const op = "bsr.(Validate).ValidateContainer"
+
+	containerValidation := &ContainerValidation{
+		Name:          name,
+		ContainerType: ct,
+	}
+
+	containerChecksumValidation, err := c.ValidateChecksums(ctx)
+	if err != nil {
+		v.Valid = false
+		containerValidation.Error = fmt.Errorf("%s: failed to validate %s: %w", op, name, err)
+		return containerValidation
+	}
+
+	containerValidation.FileChecksumValidations = containerChecksumValidation
+
+	v.updateValidationStatus(ctx, containerValidation)
+
+	return containerValidation
+}
+
+// updateValidationSummary updates the value of "Valid" field in Validation struct based on if there are failed checksums
+func (v *Validation) updateValidationStatus(ctx context.Context, cv *ContainerValidation) {
+	failedChecksums := cv.FileChecksumValidations.GetFailedItems()
+
+	// Update validation filed only if there are failed checksums
+	if len(failedChecksums) > 0 {
+		v.Valid = false
+	}
+}
```

Comment:
I don't think these should be methods on the Validation struct. In particular the ValidateContainer method being an exported method makes it seems like it could be called from outside this package, but given how it depends on the unexported type `container` it would probably be difficult to do so, and should probably only ever be called by `Validate` anyway.

I think instead, these could be unexported package functions. And the `updateValidationStatus` function could probably just be included in the other function, since it is only called by that function and is rather trivial.